### PR TITLE
[ET-1349] Add duplicate ticket feature

### PR DESCRIPTION
**Pull request by codingmusician**
### 🎫 Ticket

[ET-1349] 

### 🗒️ Description

Adds a way to duplicate a ticket in Classic Editor. It should duplicate the selected ticket and copy ticket meta to the newly created ticket.

### 🎥 Artifacts 

https://youtu.be/DnEZW84KpBY

### ✔️ Checklist
- [ ] I've included a changelog entry in the readme.txt file. 
- [x] My code is tested. 
- [x] My code has [proper inline documentation](https://the-events-calendar.github.io/products-engineering/docs/code-standards/).

### CODE REVIEW

File: src/Tribe/Metabox.php [hunk 1]
```diff
@@ -345,6 +346,123 @@ public function ajax_ticket_delete() {
...
+		$post_id = absint( tribe_get_request_var( 'post_id', 0 ) );
+
+		if ( ! $post_id ) {
+			wp_send_json_error( esc_html__( 'Invalid parent Post', 'event-tickets' ) );
+		}
+
+		$ticket_id = absint( tribe_get_request_var( 'ticket_id', 0 ) );
+
+		if ( ! $ticket_id ) {
+			wp_send_json_error( esc_html( sprintf( __( 'Invalid %s', 'event-tickets' ), tribe_get_ticket_label_singular( 'ajax_ticket_duplicate_error' ) ) ) );
```
Comment (juanfra) [inline src/Tribe/Metabox.php:364]: We need to add translators comments whenever we use sprint/printf.

File: src/Tribe/Metabox.php [hunk 2]
```diff
@@ -345,6 +346,123 @@ public function ajax_ticket_delete() {
...
+		}
+
+		if ( ! $this->has_permission( $post_id, $_POST, 'duplicate_ticket_nonce' ) ) {
+			wp_send_json_error( esc_html( sprintf( __( 'Failed to duplicate the %s. Refresh the page to try again.', 'event-tickets' ), tribe_get_ticket_label_singular( 'ajax_ticket_duplicate_error' ) ) ) );
```
Comment (juanfra) [inline src/Tribe/Metabox.php:368]: We need to add translators comments whenever we use sprint/printf.

File: src/Tribe/Metabox.php [hunk 3]
```diff
@@ -345,6 +346,123 @@ public function ajax_ticket_delete() {
...
+			return new WP_Error(
+				'bad_request',
+				__( 'Ticket ID invalid', 'event-tickets' ),
+				[ 'status' => 400 ]
+			);
+		}
+		
+		// Create data for duplicate ticket.
+		$data = [
+			'ticket_name'             => $ticket->name . __( ' (duplicate)', 'event-tickets'),
```
Comment (juanfra) [inline src/Tribe/Metabox.php:394]: Maybe we can use ` (copy)` and move the space out of the string?

Comment (juanfra) [inline src/Tribe/Metabox.php:394]:
```suggestion
			'ticket_name'             => $ticket->name . __( ' (duplicate)', 'event-tickets' ),
```

File: src/Tribe/Metabox.php [hunk 4]
```diff
@@ -345,6 +346,123 @@ public function ajax_ticket_delete() {
...
+				'_wp_old_slug',
+				'total_sales',
+			];
+			$ticket_meta = get_post_meta( $ticket->ID );
+			
+			if ( $ticket_meta ) {
+				foreach ( $ticket_meta as $meta_key => $meta_values ) {
+					
+					// Skip meta we don't want to duplicate.
+					if ( strpos( $meta_key, '_tec_tc_ticket_status_count' ) !== false ){
```
Comment (juanfra) [inline src/Tribe/Metabox.php:427]:
```suggestion
					if ( false !== strpos( $meta_key, '_tec_tc_ticket_status_count' ) ) {
```
We need to use yoda conditional here.

File: src/Tribe/Metabox.php [hunk 5]
```diff
@@ -345,6 +346,123 @@ public function ajax_ticket_delete() {
...
+		$return = $this->get_panels( $post_id );
+		$return['notice'] = $this->notice( 'ticket-duplicate' );
+
+		/**
+		 * Filters the return data for ticket duplicate.
+		 *
+		 * @param array $return  Array of data to return to the ajax call.
+		 * @param int   $post_id ID of parent "event" post.
+		 */
+		$return = apply_filters( 'event_tickets_ajax_ticket_duplicate_data', $return, $post_id );
```
Comment (juanfra) [inline src/Tribe/Metabox.php:461]: Should we also include the new ticket id as part of the params?

File: src/Tribe/Metabox.php [hunk 6]
```diff
@@ -345,6 +346,123 @@ public function ajax_ticket_delete() {
...
+		} else {
+			wp_send_json_error( esc_html( sprintf( __( 'Failed to duplicate the %s', 'event-tickets' ), tribe_get_ticket_label_singular( 'ajax_ticket_duplicate_error' ) ) ) );
+		}
+
+		$return = $this->get_panels( $post_id );
+		$return['notice'] = $this->notice( 'ticket-duplicate' );
+
+		/**
+		 * Filters the return data for ticket duplicate.
+		 *
```
Comment (juanfra) [inline src/Tribe/Metabox.php:457]: We need to add a `@since TBD` here.

File: src/Tribe/Metabox.php [hunk 7]
```diff
@@ -345,6 +346,123 @@ public function ajax_ticket_delete() {
...
+				'capacity' => $ticket->capacity(),
+				'mode'     => $ticket->global_stock_mode(),
+			]
+		];
+
+		// Add the ticket.
+		$duplicate_ticket_id = $provider->ticket_add( $post_id, $data );
+
+		// Successful?
+		if ( $duplicate_ticket_id ) {
```
Comment (juanfra) [inline src/Tribe/Metabox.php:412]: Would it be better to move the contents of this if to a new method?

File: src/admin-views/editor/list-row.php [hunk 8]
```diff
@@ -108,6 +108,18 @@ class="tribe-ticket-field-order"
 
 	<td class="ticket_edit">
 		<?php
+		printf(
+			"<button data-provider='%s' data-ticket-id='%s' title='%s' class='ticket_duplicate'><span class='ticket_duplicate_text'>%s</span></a>",
+			esc_attr( $ticket->provider_class ),
+			esc_attr( $ticket->ID ),
+			esc_attr( sprintf(
+				_x( 'Duplicate %s ID: %d', 'ticket ID title attribute', 'event-tickets' ),
```
Comment (juanfra) [inline src/admin-views/editor/list-row.php:116]: We need to add translators comments whenever we use sprint/printf.

**Review by juanfra — COMMENTED** [review]
This is looking awesome 🎉

File: src/Tribe/Metabox.php (hunk 5 above)
Comment (codingmusician) [reply src/Tribe/Metabox.php:461]: Not necessarily, since it is specifically filtering the returned data from the `get_panels()` method, which only accepts the `$post_id`.

File: src/Tribe/Metabox.php (hunk 7 above)
Comment (codingmusician) [reply src/Tribe/Metabox.php:420]: Would you recommend adding a `Tribe__Tickets__Tickets::ticket_duplicate()` method, maybe?

Comment (juanfra) [reply src/Tribe/Metabox.php:420]: Exactly, something like that. Something that can help with readability and clarity. When methods get too long is hard to follow them (sometimes) and for this particular case I feel we can move this piece out.

File: src/Tribe/Tickets.php [hunk 9]
```diff
@@ -3623,6 +3623,77 @@ public function get_ticket_form_hook() {
 
 			return $ticket_form_hook;
 		}
+		
+		/**
+		 * Creates a duplicate ticket based on post id and ticket id.
+		 *
```
Comment (juanfra) [inline src/Tribe/Tickets.php:3629]:
```suggestion
		 *
		 * @since TBD
		 *
```

**Review by juanfra — DISMISSED** [review]

**Review by juanfra — APPROVED** [review]

**Review by rafsuntaskin — DISMISSED** [review]

**Review by rafsuntaskin — APPROVED** [review]

**Review by juanfra — APPROVED** [review]
